File: modules/typed-ids-jackson2/src/main/java/org/framefork/typedIds/uuid/json/jackson/ObjectUuidSerializer.java
package org.framefork.typedIds.uuid.json.jackson;

import com.fasterxml.jackson.core.JsonGenerator;
import com.fasterxml.jackson.databind.JsonSerializer;
import com.fasterxml.jackson.databind.SerializerProvider;
import org.framefork.typedIds.uuid.ObjectUuid;

import java.io.IOException;

public class ObjectUuidSerializer extends JsonSerializer<ObjectUuid<?>>
{

    @Override
    public void serialize(
        final ObjectUuid<?> uuid,
        final JsonGenerator jsonGenerator,
        final SerializerProvider serializerProvider
    ) throws
        IOException
    {
        jsonGenerator.writeString(uuid.toNativeUuid().toString());
    }

}

File: modules/typed-ids/src/main/kotlin/org/framefork/typedIds/uuid/json/kotlinxSerialization/ObjectUuidSerializer.kt
package org.framefork.typedIds.uuid.json.kotlinxSerialization

import kotlinx.serialization.KSerializer
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import org.framefork.typedIds.uuid.ObjectUuid
import java.util.*

open class ObjectUuidSerializer<T : ObjectUuid<*>>(
    private val fromUuidConstructor: (UUID) -> T,
) : KSerializer<T> {

    override val descriptor: SerialDescriptor = PrimitiveSerialDescriptor("ObjectUuid", PrimitiveKind.STRING)

    override fun serialize(encoder: Encoder, value: T) {
        encoder.encodeString(value.toString());
    }

    override fun deserialize(decoder: Decoder): T {
        val rawUuid = UUID.fromString(decoder.decodeString())
        return fromUuidConstructor.invoke(rawUuid)
    }

}
